src/common/btree.c

```diff
@@ -1161,19 +1174,99 @@ btr_node_split_and_insert(struct btr_context *tcx, struct btr_trace *trace,
 	return rc;
 }
 
+static inline bool
+btr_root_resize_needed(struct btr_context *tcx)
+{
+	struct btr_root	*root = tcx->tc_tins.ti_root;
+	struct btr_node *nd;
+
+	if (tcx->tc_order == root->tr_node_size)
+		return false; /* As big as it can get */
+
+	nd = btr_off2ptr(tcx, root->tr_node);
+
+	if (nd->tn_keyn != root->tr_node_size)
+		return false;
+
+	return true;
+}
+
+static int
+btr_root_resize(struct btr_context *tcx, struct btr_trace *trace,
+		bool *node_alloc)
+{
+	struct btr_root	*root = tcx->tc_tins.ti_root;
+	umem_off_t	 old_node = root->tr_node;
+	struct btr_node	*nd = btr_off2ptr(tcx, old_node);
+	daos_size_t	 old_size = btr_node_size(tcx);
+	int		 new_order;
+	umem_off_t	 nd_off;
+	int		 rc = 0;
+
+	D_ASSERT(root->tr_depth == 1);
+
+	if (btr_has_tx(tcx)) {
+		rc = btr_root_tx_add(tcx);
+		if (rc != 0) {
+			D_ERROR("Failed to add btr_root to transaction\n");
+			return rc;
+		}
+	}
+
+	root->tr_node_size_idx++;
+	new_order = MIN(btree_node_size_idx[root->tr_node_size_idx],
+			tcx->tc_order);
```

Comment:
I think we don't have to store this size_idx at all, instead we could simply have 
```new_order = MIN(root->tr_node_size * 2 + 1, tcx->tc_order)```

Reply:
Yeah, seems reasonable.